File: pilhas-filas/javascript/is-balanced-stack2.js
import Stack2 from "./stack2.js";

function isBalanced(text) {
  let stack = new Stack2();
  for (let char of text) {
    if (char === "(") {
      stack.push(char);
    } else if (char === ")") {
      if (stack.isEmpty()) {
        return false;
      }
      stack.pop();
    }
  }
  return stack.isEmpty();
}

console.log(isBalanced("(())()")); // true
console.log(isBalanced("())(")); // false

File: pilhas-filas/java/src/IsBalancedStack2.java
public class IsBalancedStack2 {
	public static boolean isBalanced(String text) {
        Stack2<Character> stack = new Stack2<>();
        for (char ch : text.toCharArray()) {
            if (ch == '(') {
                stack.push(ch);
            } else if (ch == ')') {
                if (stack.isEmpty()) {
                    return false;
                }
                stack.pop();
            }
        }
        return stack.isEmpty();
    }

    public static void main(String[] args) {
        System.out.println(isBalanced("(())()")); // true
        System.out.println(isBalanced("())("));   // false
    }
}
